### PR TITLE
Extend the discussion in our exception module.

### LINKED ISSUE #4038: Properly document the purpose of Assert, AssertThrow, AssertNothrow
Reading over the description of the three macros I realize that we didn't do a stellar job in explaining the purpose of our three variants of asserts. We should improve the documentation. Further, we should check:
 - whether an `AssertThrow` should really be an `AssertThrow`, or whether `Assert` is more appropriate.
 - whether we correctly guard all asserts in destructors with `AssertNothrow` (that cannot throw).

In short:
 - `Assert` is our main assert macro that issues an error and aborts the program (normal behavior). This is the most useful macro for debugging, because a simple backtrack of the abort leads to the correct location in source code.
 - `AssertThrow` - sometimes input depends on runtime supplied user data. in this case using the C++ exception mechanism is more appropriate. (A program can catch an exception at runtime and correct for it). This macro must not be used for internal consistency checks, because it makes debugging considerably harder.
 - `AssertNothrow` - this macro is the same as `Assert` except that it guarantees to never throw an exception. This is needed for asserts in destructors (throwing an exception in a destructor is undefined behavior).

### PR DESCRIPTION
**Pull request by bangerth**
This fixes #4038.

### CODE REVIEW

File: doc/doxygen/headers/exceptions.h [hunk 1]
```diff
@@ -351,5 +381,49 @@
...
+ *
+ *  The default implementation of the @p Assert macro, as discussed above,
+ *  prints detailed information about what exactly went wrong to the
+ *  screen and then aborts the program. Aborting the program is useful
+ *  because it allows easily finding the place where something went
+ *  wrong -- including all of the information how we got to that
+ *  place -- by running the program in a debugger.
+ *
+ *  On the other hand, there are cases where aborting a program may be
+ *  undesirable and one needs to exit in a somewhat more gracefully
```
Comment (kronbichler) [inline doc/doxygen/headers/exceptions.h]: gracefully -> graceful

File: doc/doxygen/headers/exceptions.h [hunk 2]
```diff
@@ -351,5 +381,49 @@
...
+ *  same as @p AssertThrow does, namely using the C++ @p throw mechanism
+ *  to raise an exception. This exception can then be caught at a higher
+ *  level -- e.g., in the optimization routine that sits atop the flow
+ *  solver, and that can then decide what it wants to do with the
+ *  situation.
+ *
+ *  This is all nice and good, but C++ does not allow throwing exceptions
+ *  inside the destructors of classes, or in a function that is currently
+ *  being called from a destructor higher up in the call stack. To this
+ *  end, there is separate macro, @p AssertNothrow, that can be used in
```
Comment (kronbichler) [inline doc/doxygen/headers/exceptions.h]: there is A separate macro

File: doc/doxygen/headers/exceptions.h [hunk 3]
```diff
@@ -351,5 +381,49 @@
...
+ *  destructors: It acts just like @p Assert usually does -- in particular,
+ *  in only checks the condition in debug mode -- but it is immune to the
```
Comment (kronbichler) [inline doc/doxygen/headers/exceptions.h]: IT only checks...

**Comment by bangerth** [comment]
Thanks for the review, @kronbichler . I fixed the places you marked up!

**Pull request merged**